lib/writeUris.js

```diff
@@ -0,0 +1,126 @@
+'use strict';
+var fsExtra = require('fs-extra');
+var mime = require('mime');
+var path = require('path');
+var Promise = require('bluebird');
+
+var fsExtraOutputFile = Promise.promisify(fsExtra.outputFile);
+
+module.exports = writeUris;
+
+/**
+ * Write glTF resources as embedded data uris or external files.
+ *
+ * @param {Object} gltf The glTF asset.
+ * @param {String} gltfPath Path where the glTF will be saved.
+ * @param {Boolean} separateBuffers Writes out separate buffers.
+ * @param {Boolean} separateTextures Writes out separate textures.
+ * @returns {Promise} A promise that resolves to the glTF asset.
+ *
+ * @private
+ */
+function writeUris(gltf, gltfPath, separateBuffers, separateTextures) {
+    var promises = [];
+
+    var buffer = gltf.buffers[Object.keys(gltf.buffers)[0]];
+    var bufferByteLength = buffer.extras._obj2gltf.source.length;
+
+    var texturesByteLength = 0;
+    var images = gltf.images;
+    for (var id in images) {
+        if (images.hasOwnProperty(id)) {
+            texturesByteLength += images[id].extras._obj2gltf.source.length;
+        }
+    }
+
+    // Buffers larger than ~192MB cannot be base64 encoded due to a NodeJS limitation. Source: https://github.com/nodejs/node/issues/4266
+    var exceedsMaximum = (texturesByteLength + bufferByteLength > 201326580);
+
+    if (exceedsMaximum) {
+        console.log('Buffers and textures are too large to encode in the glTF, saving as separate resources.');
+    }
+
+    if (separateBuffers || exceedsMaximum) {
+        promises.push(writeSeparateBuffer(gltf, gltfPath));
+    } else {
+        writeEmbeddedBuffer(gltf);
+    }
+
+    if (separateTextures || exceedsMaximum) {
+        promises.push(writeSeparateTextures(gltf, gltfPath));
+    } else {
+        writeEmbeddedTextures(gltf);
+    }
+
+    deleteExtras(gltf);
```

Comment:
I wonder if it's possible for this to get called before `writeSeparateX` finishes on something like a model with a huge number of textures. It might be safer to put this right before the `return gltf;` below.